src/cycles/simple_cycles.c

```diff
@@ -0,0 +1,506 @@
+                        IGRAPH_FINALLY_CLEAN(2);
+                    }
+                }
+                igraph_vector_int_pop_back(&state->edge_stack);
+            } else if (!(VECTOR(state->blocked)[W])) {
+                // printf("Recursing deeper from %" IGRAPH_PRId " to  %" IGRAPH_PRId "\n", V, W);
+
+                recurse_deeper = true;
+                IGRAPH_CHECK(igraph_stack_int_push(&neigh_iteration_progress, i + 1));
+                IGRAPH_CHECK(igraph_stack_int_push(&v_stack, V));
+                IGRAPH_CHECK(igraph_stack_int_push(&e_stack, E));
+                V = W;
+                E = WE;
+                break;
+            }
+        }
+
+        if (!recurse_deeper) {
+            // L2
+            if (local_found) {
+                // IGRAPH_CHECK(igraph_i_simple_cycles_unblock_recursive(state, V));
+                IGRAPH_CHECK(igraph_i_simple_cycles_unblock(state, V));
+            } else {
+                for (igraph_integer_t i = 0; i < num_neighbors; ++i) {
+                    igraph_integer_t W = VECTOR(*neighbors)[i];
+                    igraph_integer_t pos;
+                    if (!igraph_vector_int_search(igraph_adjlist_get(&state->B, W), 0, V, &pos)) {
+                        IGRAPH_CHECK(igraph_vector_int_push_back(igraph_adjlist_get(&state->B, W), V));
+                    }
+                }
+            }
+
+            IGRAPH_ASSERT(!igraph_vector_int_empty(&state->vertex_stack));
+
+            // unstack v
+            igraph_vector_int_pop_back(&state->vertex_stack);
+            if (!igraph_vector_int_empty(&state->edge_stack)) {
+                // can be empty for the starting point.
+                // alternatively, V == S
+                E = igraph_vector_int_pop_back(&state->edge_stack);
+            }
+        }
+    }
+
+    *found = local_found;
+
+    IGRAPH_ASSERT(igraph_stack_int_size(&v_stack) == 0);
+    IGRAPH_ASSERT(igraph_stack_int_size(&e_stack) == 0);
+    IGRAPH_ASSERT(igraph_stack_int_size(&neigh_iteration_progress) == 0);
+    igraph_stack_int_destroy(&neigh_iteration_progress);
+    igraph_stack_int_destroy(&v_stack);
+    igraph_stack_int_destroy(&e_stack);
+    IGRAPH_FINALLY_CLEAN(3);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_simple_cycle_search_state_init
+ * \brief Initialize the cycle search state
+ *
+ * \experimental
+ *
+ * \param state The state structure to initialize
+ * \param graph The graph object
+ * \return Error code.
+ *
+ * Time complexity: O(|V|*|E|*log(|V|*|E|))
+ *
+ * \ref igraph_simple_cycle_search_state_destroy
+ */
+igraph_error_t igraph_simple_cycle_search_state_init(
+    igraph_simple_cycle_search_state_t *state, const igraph_t *graph) {
+    state->N = igraph_vcount(graph);
+
+    IGRAPH_CHECK(igraph_vector_int_init(&state->vertex_stack, 0));
+    IGRAPH_CHECK(igraph_vector_int_reserve(&state->vertex_stack, 8));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &state->vertex_stack);
```

Comment:
The pattern here is slight incorrect. The `IGRAPH_FINALLY` should come directly after the `init()`, and before the `reserve()`. We can combine the `init` and `FINALLY` into a single `IGRAPH_VECTOR_INT_INIT_FINALLY`.

```suggestion
    IGRAPH_VECTOR_INT_INIT_FINALLY(&state->vertex_stack, 0);
    IGRAPH_CHECK(igraph_vector_int_reserve(&state->vertex_stack, 8));
```

Before merging these will need to be cleaned up.